src/room/room-prototype.ts

```diff
@@ -63,18 +64,93 @@ const findNextEnergySource = function(creep: Creep) {
             }
         }
     }
-}
+};
+
+const getNumberOfMiningSpaces = function() {
+    if (!this.memory['sources'] || !this.memory['sources']['spaces']) {
+        this.findNumberOfSourcesAndSpaces();
+    }
+    return this.memory['sources']['spaces'];
+};
+
+const getNumberOfSources = function() {
+    if (!this.memory['sources'] || !this.memory['sources']['count']) {
+        this.findNumberOfSourcesAndSpaces();
+    }
+    return this.memory['sources']['count'];
+};
+
+const findNumberOfSourcesAndSpaces = function() {
+    let numberSources = 0;
+    let numberSpaces = 0;
+    _.forEach(this.find(FIND_SOURCES), (source: Source) => {
+        numberSources++;
+        const availablePositions = {};
```

Comment:
`lookAtArea` returns multiple `LookAtResultWithPos` for the same coord. You could have a swamp, creep, road, and rampart all on the same spot and it would return 4 results for that coord. Using the hashset `availablePositions` I can track which positions haven't been ruled out.